mm2src/coins/solana.rs

```diff
@@ -378,6 +396,246 @@ impl SolanaCoin {
+            AccountMeta::new(vault_pda_data, false),        // Not a signer
+            AccountMeta::new(vault_pda, false),             // Not a signer
+            AccountMeta::new(solana_program::system_program::id(), false), //system_program must be included
+        ];
+        self.sign_and_send_transaction(swap_program_id, accounts, swap_instruction.pack())
+    }
+
+    fn get_transaction_details(&self, tx_hash: &[u8]) -> Result<(u64, [u8; 32], u64, Pubkey), Box<TransactionErr>> {
+        let transaction = deserialize(tx_hash);
+        let transaction: SolTransaction =
+            transaction.map_err(|e| Box::new(TransactionErr::Plain(ERRL!("error deserializing tx_hash: {:?}", e))))?;
+
+        if let Some(instruction) = &transaction.message.instructions.get(0) {
+            let data = &instruction.data;
+            let instruction_data = &data[..];
+            let instruction = AtomicSwapInstruction::unpack(instruction_data[0], instruction_data)
+                .map_err(|e| Box::new(TransactionErr::Plain(ERRL!("error unpacking tx data: {:?}", e))))?;
+
+            match instruction {
+                AtomicSwapInstruction::LamportsPayment {
+                    secret_hash,
+                    lock_time,
+                    amount,
+                    receiver,
+                    rent_exemption_lamports,
+                    vault_bump_seed,
+                    vault_bump_seed_data,
+                } => Ok((lock_time, secret_hash, amount, Pubkey::new_from_array([0; 32]))),
+                AtomicSwapInstruction::SLPTokenPayment {
+                    secret_hash,
+                    lock_time,
+                    amount,
+                    receiver,
+                    token_program,
+                    rent_exemption_lamports,
+                    vault_bump_seed,
+                    vault_bump_seed_data,
+                } => Ok((lock_time, secret_hash, amount, token_program)),
+                AtomicSwapInstruction::ReceiverSpend {
+                    secret,
+                    lock_time,
+                    amount,
+                    sender,
+                    token_program,
+                    vault_bump_seed,
+                    vault_bump_seed_data,
+                } => Ok((lock_time, sha256(&secret).take(), amount, token_program)),
+                AtomicSwapInstruction::SenderRefund {
+                    secret_hash,
+                    lock_time,
+                    amount,
+                    receiver,
+                    token_program,
+                    vault_bump_seed,
+                    vault_bump_seed_data,
+                } => Ok((lock_time, secret_hash, amount, token_program)),
+            }
+        } else {
+            Err(Box::new(TransactionErr::Plain(ERRL!(
+                "Instruction not found in message"
+            ))))
+        }
+    }
+
+    pub fn sign_and_send_transaction(&self, program_id: Pubkey, accounts: Vec<AccountMeta>, data: Vec<u8>) -> SolTxFut {
```

Comment:
This function can be made async
```rust
    pub async fn sign_and_send_swap_transaction(
        &self,
        program_id: Pubkey,
        accounts: Vec<AccountMeta>,
        data: Vec<u8>,
    ) -> Result<SolTransaction, TransactionErr> {
        // Construct the instruction to send to the program
        // The parameters here depend on your specific program's requirements
        let instruction = Instruction {
            program_id,
            accounts, // Specify account metas here
            data,     // Pass data to the program here
        };

        // Create a transaction
        let recent_blockhash = self
            .client
            .get_latest_blockhash()
            .map_err(|e| TransactionErr::Plain(format!("Failed to get recent blockhash: {:?}", e)))?;

        let transaction: SolTransaction = SolTransaction::new_signed_with_payer(
            &[instruction],
            Some(&self.key_pair.pubkey()), //payer pubkey
            &[&self.key_pair],             //payer
            recent_blockhash,
        );

        // Send the transaction
        let tx = self.client
            .send_and_confirm_transaction(&transaction)
            .map(|_signature| transaction)
            .map_err(|e| TransactionErr::Plain(ERRL!("Solana ClientError: {:?}", e)))?;

        Ok(tx)
    }
```
and used in other places that return a future as such
```rust
        let fut = self.sign_and_send_transaction(swap_program_id, accounts, swap_instruction.pack());
        Box::new(fut.boxed().compat())
```